### PR TITLE
sql: refactor type types

### LINKED ISSUE #10571: sql: refactor type types
We've got a lot of confusion around our scalar types at the moment, between `repr::ScalarType`, `pgrepr::Type`, `coord::catalog::Type`, and `postgres_types::Type`. Here's a series of various refactorings I have in mind:

  * The abbreviated term "typmod" is used exclusively to refer to a PostgreSQL-style packed 32-bit integer describing how to modify a base type.
  * The term "modifiers" is used in the SQL parser to refer to a list of arbitrary `Value`s attached to the type. These are _unpacked_ type modifiers.
  * The code in `repr` usually speaks very concretely about e.g. the "max scale" of a numeric or the "max length" of a varchar. It may occasionally refer to this class of thing as "type modifiers", but it will be careful to avoid the shorthand "typmod".
  * We introduce wrapper types called `NumericMaxScale` and `VarCharMaxLength` and such that validate the bounds of the provided integer. These wrapper types offer infallible conversions to usize and i32 and whatever other types are convenient, and offer a _fallible_ conversion from other numeric types. 
  * The `extract_typ_mods` family of functions moves out `repr` and into `sql`. This code is exclusively used in the SQL parser -> `ScalarType` conversion, and so belongs in the planner, not `repr`.
  * The catalog exposes the `TypeInner` struct to the `sql` crate, so that the problem of converting a SQL parser data type into a `ScalarType` can live entirely in the SQL planner, rather than in a `try_get_lossy_scalar_type_by_id` method that lives in the catalog.

### LINKED ISSUE #10576: repr: create newtypes for `NumericMaxScale` and `[Var]CharMaxLength`
See #10571 for context. It's a bit scary that it's possible to construct a `ScalarType::Numeric` or `ScalarType::[Var]Char` with invalid scale/length fields. We should introduce newtypes for these fields that perform the validation on construction.

### PR DESCRIPTION
**Pull request by benesch**
This mega PR refactors the type types as described in #10571. More details in that issue and in the commit messages within. The net result is, IMO, a set of types whose purposes are more clear, safer and more correct conversions between them, and the removal of a *lot* of duplicate code. 

I'm sorry this diff is so large, but these changes were massively interdependent, and I don't have any better ideas for further splitting it into chunks.

Fix #10571.
Fix #10576.

### Motivation

   * This PR refactors existing code.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/pgrepr/src/types.rs [hunk 1]
```diff
@@ -84,6 +98,89 @@ pub enum Type {
...
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct NumericConstraints {
+    /// The maximum precision.
+    max_precision: i32,
+    /// The maximum scale.
+    max_scale: i32,
+}
+
+impl NumericConstraints {
+    fn from_typmod(typmod: i32) -> Option<NumericConstraints> {
```
Comment (sploiselle) [inline src/pgrepr/src/types.rs:159]: 🤯 This seems like wants to share an implementation w/ https://github.com/MaterializeInc/materialize/blob/e49579ca8dd8e227e13bfd58df707d1ccf0e07b3/src/expr/src/scalar/func.rs#L5691, though maybe I'm being overeager

Once this PR lands, will see if I can find a place to tie these together.

File: src/postgres-util/src/lib.rs [hunk 2]
```diff
@@ -9,127 +9,23 @@
...
-pub struct PgNumericMod {
-    precision: u16,
-    scale: u16,
-}
-
-/// Mirror of PostgreSQL's
-/// [`VARHDRSZ`](https://github.com/postgres/postgres/blob/REL_14_0/src/include/c.h#L627) constant
-const PG_HEADER_SIZE: i32 = 4;
-
-pub enum PgScalarType {
```
Comment (sploiselle) [inline src/postgres-util/src/lib.rs:36]: 🙇

**Review by sploiselle — APPROVED** [review]
I obviously didn't spend time reverse-engineering the motivation behind all of these changes. However, I "spot checked" all of the major points of refactoring and things seem really great/clear.

If you have the time or interest, I'd love to hear your high-level thinking behind:
- `PgScalarType` being the unnecessary pivot between `pgrepr` and `ScalarType`. I believe you mentioned removing this to me previously, but did you know it was chaff because you had to write it originally, or was this something you discovered by means of "looking at"/understanding the dependency graph?
- New typing the typmods in `adt`. Like, is it nice to tidy up the semantic boundaries of these things? I saw a "greatest common denominator" for all of them, so never would have been motivated to specialize the code. Would love to get a clearer sense of why you think this approach is superior (asking to learn, not in any way a disagreement).

**Review by nharring-adjacent — APPROVED** [review]
This all looks reasonable, I mostly focused on the postgres_util and pgrepr bits since I'm more familiar with that code however everything is clear and readable. Thanks very much for doing this, these type relationships seem demonstrably easier to grok and use.

**Comment by benesch** [comment]
> `PgScalarType` being the unnecessary pivot between pgrepr and ScalarType. I believe you mentioned removing this to me previously, but did you know it was chaff because you had to write it originally, or was this something you discovered by means of "looking at"/understanding the dependency graph?

A combination of both, I think! When I wrote `pgrepr::Type` I meant for it to be an exact representation of _Postgres's_ view of the type. So seeing a type called `PgScalarType` pop up definitely triggered my spidey sense. But `pgrepr::Type` wasn't directly usable as `PgScalarType` for sure, because it had accumulated some cruft over the years and didn't handle typmods directly.

> New typing the typmods in adt. Like, is it nice to tidy up the semantic boundaries of these things? I saw a "greatest common denominator" for all of them, so never would have been motivated to specialize the code. Would love to get a clearer sense of why you think this approach is superior (asking to learn, not in any way a disagreement).

Oh, I just get stressed out about having representable states that are invalid. Like, nothing stopped someone from inadvertently constructing `ScalarType::Numeric { scale: Some(400) }`, and falling into completely unexercised code paths. This doesn't matter _too_ much when you're only constructing `ScalarType::Numeric` in one place, but since there's a few different ways of getting there now (via the SQL parser -> planner, via `pgrepr`, via the catalog, and via `interchange`), it just seemed worth completely eliminating the possibility of error. I can now say with confidence that nothing in our codebase can construct a `ScalarType::Numeric` with a scale that exceeds 38. On the other hand: still not convinced that everything that constructs a decimal is properly restricting the precision to 38!

File: src/pgrepr/src/types.rs (hunk 1 above)
Comment (benesch) [reply src/pgrepr/src/types.rs:159]: Oh 100% it should!! Good catch!

**Comment by benesch** [comment]
I had to ditch the last commit that updated the test that compares PostgreSQL OIDs to Materialize OIDs, because it was failing in CI for reasons entirely unrelated to this commit. I'll do that in a separate PR to get this unblocked.

**Comment by benesch** [comment]
Hallelujah! Thanks for the reviews, everyone!

**Pull request merged**